cupy/lib/polynomial.pyx

```diff
@@ -0,0 +1,275 @@
+import re
+import warnings
+
+import numpy
+
+import cupy
+
+_poly_mat = re.compile(r'[*][*]([0-9]*)')
+
+
+def _raise_power(astr, wrap=70):
+    n = 0
+    line1 = ''
+    line2 = ''
+    output = ' '
+    while True:
+        mat = _poly_mat.search(astr, n)
+        if mat is None:
+            break
+        span = mat.span()
+        power = mat.groups()[0]
+        partstr = astr[n:span[0]]
+        n = span[1]
+        toadd2 = partstr + ' ' * (len(power) - 1)
+        toadd1 = ' ' * (len(partstr) - 1) + power
+        if ((len(line2) + len(toadd2) > wrap) or
+                (len(line1) + len(toadd1) > wrap)):
+            output += line1 + '\n' + line2 + '\n '
+            line1 = toadd1
+            line2 = toadd2
+        else:
+            line2 += partstr + ' ' * (len(power) - 1)
+            line1 += ' ' * (len(partstr) - 1) + power
+    output += line1 + '\n' + line2
+    return output + astr[n:]
+
+
+class poly1d(object):
+    """A one-dimensional polynomial class.
+
+    Args:
+        c_or_r (cupy.ndarray or cupy.poly1d): The polynomial's
+         coefficients in decreasing powers
+        r (bool, optional): If True, `c_or_r` specifies the
+            polynomial's roots; the default is False.
+        variable (str, optional): Changes the variable used when
+            printing the polynomial from `x` to `variable`
+
+    .. seealso:: :func:`numpy.poly1d`
+
+    """
+    __hash__ = None
+
+    @property
+    def coeffs(self):
+        return self._coeffs
+
+    @coeffs.setter
+    def coeffs(self, value):
+        if value is not self._coeffs:
+            raise AttributeError('Cannot set attribute')
+
+    @property
+    def variable(self):
+        return self._variable
+
+    @property
+    def order(self):
+        return self._coeffs.size - 1
+
+    # TODO
+    @property
+    def roots(self):
+        pass
+
+    @property
+    def _coeffs(self):
+        return self.__dict__['coeffs']
+
+    @_coeffs.setter
+    def _coeffs(self, coeffs):
+        self.__dict__['coeffs'] = coeffs
```

Comment:
Why are they needed?

Reply:
According to Scipy
>  our internal _coeffs property need to be backed by __dict__['coeffs'] for scipy to work correctly.

Reply:
SciPy seems to have their original `scipy.poly1d` version, but it is just an alias of `numpy.poly1d` in recent versions. So we don't need this hack.